cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,337 @@
+    """
+    combined: 'cirq.PauliSum' = sum(hamiltonian_polynomial_list, _Hamiltonian_O())
+
+    qubit_names = sorted(qubit_map.keys())
+    qubits = [qubit_map[name] for name in qubit_names]
+    qubit_indices = {qubit: i for i, qubit in enumerate(qubits)}
+
+    hamiltonians = {}
+    for pauli_string in combined:
+        w = pauli_string.coefficient.real
+        qubit_idx = tuple(sorted(qubit_indices[qubit] for qubit in pauli_string.qubits))
+        hamiltonians[qubit_idx] = w
+
+    # Here we follow improvements of [4] cancelling out the CNOTs. The first step is to order by
+    # Gray code so that as few as possible gates are changed.
+    sorted_hs = sorted(list(hamiltonians.keys()), key=functools.cmp_to_key(_gray_code_comparator))
+
+    def _simplify_cnots(cnots):
+        _control = 0
+        _target = 1
+
+        while True:
+            # As per equations 9 and 10 of [4], if all the targets (resp. controls) are the same,
+            # the cnots commute. Further, if the control (resp. targets) are the same, the cnots
+            # can be simplified away:
+            # CNOT(i, j) @ CNOT(k, j) = CNOT(k, j) @ CNOT(i, j)
+            # CNOT(i, k) @ CNOT(i, j) = CNOT(i, j) @ CNOT(i, k)
+            found_simplification = False
+            for x, y in [(_control, _target), (_target, _control)]:
+                i = 0
+                qubit_to_index: Dict[int, int] = {}
+                for j in range(1, len(cnots)):
+                    if cnots[i][x] != cnots[j][x]:
+                        # The targets (resp. control) don't match, so we reset the search.
+                        i = j
+                        qubit_to_index = {cnots[j][y]: j}
+                        continue
+
+                    if cnots[j][y] in qubit_to_index:
+                        k = qubit_to_index[cnots[j][y]]
+                        # The controls (resp. targets) are the same, so we can simplify away.
+                        cnots = [cnots[n] for n in range(len(cnots)) if n != j and n != k]
+                        found_simplification = True
+                        break
+
+                    qubit_to_index[cnots[j][y]] = j
+
+            if found_simplification:
+                continue
+
+            # Here we apply the simplification of equation 11. Note that by flipping the control
+            # and target qubits, we have an equally valid identity:
+            # CNOT(i, j) @ CNOT(j, k) == CNOT(j, k) @ CNOT(i, k) @ CNOT(i, j)
+            # CNOT(j, i) @ CNOT(k, j) == CNOT(k, j) @ CNOT(k, i) @ CNOT(j, i)
+            for x, y in [(_control, _target), (_target, _control)]:
+                # We investigate potential pivots sequentially.
+                for j in range(1, len(cnots) - 1):
+                    # First, we look back for as long as the targets (resp. triggers) are the same.
+                    # They all commute, so all are potential candidates for being simplified.
+                    common_A: Dict[int, int] = {}
```

Comment:
don't use capital letters in variable names

Reply:
Done.